kura/org.eclipse.kura.wire.camel/src/main/java/org/eclipse/kura/wire/camel/CamelProduce.java

```diff
@@ -46,11 +49,41 @@ protected void processReceive(final CamelContext context, final String endpointU
     @Override
     protected void deactivate() {
         super.deactivate();
+        final Lock wlock = this.rwLock.writeLock();
+        wlock.lock();
+        try {
+            closeTemplate();
+        } finally {
+            wlock.unlock();
+        }
+    }
+
+    private void checkTemplate(final CamelContext context) {
+        final Lock rlock = this.rwLock.readLock();
+        rlock.lock();
+        try {
+            templateAvailable = template != null && template.getCamelContext() == context;
```

Comment:
I wouldn't use a global/class variable here. I would just return the value. Or maybe I just don't understand the reason for this pattern. Then a bit of source doc would help ;-)

Reply:
When readLock unloack there have a race condition the other threads change templateAvailable. So i declare a volatile boolean global/class variable

Reply:
I know what you mean. This is my first version
```
public class CamelProduce extends AbstractReceiverWireComponent {

    private final ReadWriteLock rwLock = new ReentrantReadWriteLock(true);

    private FluentProducerTemplate template = null;

    @Override
    protected void processReceive(final CamelContext context, final String endpointUri, final WireEnvelope envelope)
            throws Exception {
        
        boolean templateAvailable = false;
        final Lock rlock = this.rwLock.readLock();
        rlock.lock();
        try {
            templateAvailable = isTemplateAvailable(context);
        } finally {
            rlock.unlock();
        }
        if (!templateAvailable) {
            createTemplate(context);
        }
        template //
                .withBody(envelope) //
                .to(endpointUri) //
                .asyncSend();
    }

    @Override
    protected void deactivate() {
        super.deactivate();
        final Lock wlock = this.rwLock.writeLock();
        wlock.lock();
        try {
            closeTemplate();
        } finally {
            wlock.unlock();
        }
    }

    private boolean isTemplateAvailable(final CamelContext context) {
        return template != null && template.getCamelContext() == context;
    }

    private void createTemplate(final CamelContext context) {
        final Lock wlock = this.rwLock.writeLock();
        wlock.lock();
        try {
            closeTemplate();
            template = on(context);
        } finally {
            wlock.unlock();
        }
    }

    private void closeTemplate() {
        if (template != null) {
            try {
                template.stop();
            } catch (Exception ignored) {
            } finally {
                template = null;
            }
        }
    }

}


```

Reply:
I just don't want 'call template' to be within the read lock. It will reduce the efficiency.

Reply:
I don't think it's the best way to do this. Template re-create to discard messages is acceptable.
Sending messages is not efficient within a read lock.

What do you think? @ctron 

```
public class CamelProduce extends AbstractReceiverWireComponent {

    private final ReadWriteLock rwLock = new ReentrantReadWriteLock(true);

    private FluentProducerTemplate template = null;

    @Override
    protected void processReceive(final CamelContext context, final String endpointUri, final WireEnvelope envelope)
            throws Exception {

        while (true) {
            final Lock rlock = this.rwLock.readLock();
            rlock.lock();
            try {
                if (isTemplateAvailable(context)) {
                    template //
                            .withBody(envelope) //
                            .to(endpointUri) //
                            .asyncSend();
                    return;
                }
            } finally {
                rlock.unlock();
            }

            final Lock wlock = this.rwLock.writeLock();
            wlock.lock();
            try {
                if (!isTemplateAvailable(context)) {
                    createTemplate(context);
                }
            } finally {
                wlock.unlock();
            }
        }
    }

    @Override
    protected void deactivate() {
        super.deactivate();
        final Lock wlock = this.rwLock.writeLock();
        wlock.lock();
        try {
            closeTemplate();
        } finally {
            wlock.unlock();
        }
    }

    private boolean isTemplateAvailable(final CamelContext context) {
        return template != null && template.getCamelContext() == context;
    }

    private void createTemplate(final CamelContext context) {
        closeTemplate();
        template = on(context);
    }

    private void closeTemplate() {
        if (template != null) {
            try {
                template.stop();
            } catch (Exception ignored) {
            } finally {
                template = null;
            }
        }
    }

}

```